neuralmonkey/runners/beamsearch_runner.py

```diff
@@ -1,68 +1,160 @@
+    def next_to_execute(self) -> NextExecute:
+        return (self._all_encoders,
+                {"bs_outputs": self._decoder.outputs},
+                self._next_feed)
 
-        # now backtrack
-        output_tokens = []  # type: List[str]
+    # pylint: disable=too-many-locals
+    def collect_results(self, results: List[Dict]) -> None:
+        # Recompute logits
+        # Only necessary when ensembling models
+        prev_logprobs = [res["bs_outputs"].last_search_state.prev_logprobs
+                         for res in results]
+
+        # Arithmetic mean
+        ens_logprobs = (scipy.misc.logsumexp(prev_logprobs, 0)
+                        - np.log(self._num_sessions))
+
+        # The last score is computed
+        bs_outputs = results[0]["bs_outputs"]
+        step = bs_outputs.last_dec_loop_state.step - 1
+
+        self.step += step
+        self.scores = np.append(
+            self.scores,
+            bs_outputs.last_search_step_output.scores[0:step],
+            axis=0)
+
+        self.parent_ids = np.append(
+            self.parent_ids,
+            bs_outputs.last_search_step_output.parent_ids[0:step],
+            axis=0)
+
+        self.token_ids = np.append(
+            self.token_ids,
+            bs_outputs.last_search_step_output.token_ids[0:step],
+            axis=0)
+
+        if (self._decoder.max_output_len is not None and
+                self.step > self._decoder.max_output_len):
+            self.prepare_results()
+            return
+
+        # Prepare the next feed_dict (in ensembles)
+        self._next_feed = []
+        for result in results:
+            bs_outputs = result["bs_outputs"]
+
+            search_state = bs_outputs.last_search_state._replace(
+                prev_logprobs=ens_logprobs)
+
+            fd = {self._decoder.max_steps: 1,
+                  self._decoder.search_state: search_state}
+
+            dec_feedables = bs_outputs.last_dec_loop_state
+
+            # Due to the arrays in DecoderState (prev_contexts),
+            # we have to create feed for each value separately.
+            for field in self._decoder.decoder_state._fields:
+                # We do not feed the step
+                if field == "step":
+                    continue
+                tensor = getattr(self._decoder.decoder_state, field)
+                value = getattr(dec_feedables, field)
+                if isinstance(tensor, list):
+                    for t, val in zip(tensor, value):
+                        fd.update({t: val})
+                else:
+                    fd.update({tensor: value})
+
+            self._next_feed.append(fd)
+
+        if self.token_ids.size == 0:
+            return
+
+        # We assume that we can stop decoding when all tokens
+        # in the last step were <pad>
+        if np.all(np.equal(self.token_ids[-1], PAD_TOKEN_INDEX)):
+            self.prepare_results()
+    # pylint: enable=too-many-locals
+
```

Comment:
Napsal bych tady do komentáře, že se projde pozpátku spoják, jinak to působí docela krypticky.